tools/pythonpkg/src/include/duckdb_python/pybind11/python_object_container.hpp

```diff
@@ -33,32 +27,21 @@ class PythonObjectContainer {
 		py_obj.clear();
 	}
 
-	unique_ptr<PythonGILWrapper> GetLock() {
-		return make_uniq<PythonGILWrapper>();
-	}
-
-	template <class NEW_PY_TYPE>
-	void AssignInternal(typename PythonAssignmentFunction<PY_TYPE, NEW_PY_TYPE>::assign_t lambda,
-	                    NEW_PY_TYPE &new_value, PythonGILWrapper &lock) {
-		PY_TYPE obj;
-		lambda(obj, new_value);
-		PushInternal(lock, obj);
+	void Push(py::object &&obj) {
+		py::gil_scoped_acquire gil;
+		PushInternal(std::move(obj));
 	}
 
-	void PushInternal(PythonGILWrapper &lock, PY_TYPE obj) {
-		py_obj.push_back(obj);
+	const py::object &LastAddedObject() {
+		D_ASSERT(!py_obj.empty());
+		return py_obj.back();
 	}
 
-	void Push(PY_TYPE obj) {
-		auto lock = GetLock();
-		PushInternal(*lock, std::move(obj));
-	}
-
-	const PY_TYPE *GetPointerTop() {
-		return &py_obj.back();
+private:
+	void PushInternal(py::object &&obj) {
+		py_obj.emplace_back(obj);
 	}
 
-private:
-	vector<PY_TYPE> py_obj;
+	vector<py::object> py_obj;
```

Comment:
I think it only complicates the code here to make it templated on `PY_TYPE`, in the eyes of python everything is `py::object` anyways.
The same goes for the original type, this can be generalized as py::handle and we don't need `NEW_PY_TYPE` anymore.